Add network bonding and teaming topics

tahliar opened the pull request:
### Description

Convert network bonding and teaming to new-gen smart docs. 

Now with PDFs:
[network-bond-managing-devices_en.pdf](https://github.com/SUSE/doc-modular/files/12812659/network-bond-managing-devices_en.pdf) **(Updated 2023-10-05)**
[network-team-managing-devices_en.pdf](https://github.com/SUSE/doc-modular/files/12812660/network-team-managing-devices_en.pdf) **(Updated 2023-10-05)**

For the sake of the impending localisation deadline, I'd prefer to add a good troubleshooting section later, rather than rush out something mediocre. Otherwise, this is ready for final reviews.

### Are there any relevant issues/feature requests?

* bsc#...
* jsc#SLE-...

tahliar commented:
@taroth21 I started by converting the existing smart docs into topics. I'm also planning to convert [Hotplugging of bond ports](https://documentation.suse.com/sles/15-SP4/html/SLES-all/cha-network.html#sec-network-iface-bonding-hotplug) and [Use case: VLAN over team device](https://documentation.suse.com/sles/15-SP4/html/SLES-all/cha-network.html#sec-network-iface-teaming-vlan) from the legacy docs. 

I'm still deciding whether to put both bonding and teaming into a single assembly, or put them in separate assemblies. At the moment I'm leaning towards separate, something like this: 

```
Managing network bonding devices
    Network bonding or network teaming: feature comparison
    Configuring network bonding
    Hotplugging of bond ports (probably retitle)
```
 
```
Managing network teaming devices
  Network bonding or network teaming: feature comparison
  Configuring network teaming
  Use case: VLAN over team device (retitle)
  Removing a network teaming device
```

But I'm open to suggestions if you have other ideas.

tahliar commented on articles/network-bond-managing-devices.asm.xml:
```diff
@@ -0,0 +1,144 @@
+      <meta name="updated" content="2023-06-23" its:translate="no"/>
+      <meta name="architecture" content="x86;power" its:translate="no"/>
+      <meta name="productname" its:translate="no">
+        <productname version="1">&alp;</productname>
```
@janajaeger For things like architecture and productname, what do you think about having all valid values appear in the template, and then people can delete whatever isn't relevant?

janajaeger replied:
Might be a good idea, but then again people might forget about the deletion part. I'd rather have the authors invest some thought and effort here, but that's just me ;)

tahliar replied:
Do we have a list somewhere of the possible valid values, then? I assume they need to be in a specific format and we can't just make things up...? 
Like, for example, I don't think we're including HA in smart docs yet, but if/when we do then this assembly would apply to that too.

janajaeger replied:
For productname, the standard product entities (full version, not the abbreviated ones) apply. For architecture, use anything that is currently used for profiling. HTH

tahliar replied:
Thanks, that does help, but I'm mainly suggesting having this written somewhere for future writers :)

janajaeger replied:
I'd rather not do that - we'll end  up updating that thing every other month when the ALP folks come up with new ideas.

tahliar commented on articles/network-bond-managing-devices.asm.xml:
```diff
@@ -0,0 +1,144 @@
+          <term>EFFORT</term>
+            <listitem>
+              <para>
+                What's the effort one has to put in?
```
I'm terrible at estimating time, but I could test it again and time myself...?

janajaeger replied:
Yup, that's what everybody else does. And it's not just reading time - add some time for "digestion" of the content.

tahliar replied:
Wait, so is it reading time or time to actually perform the task? Or both?

janajaeger replied:
Reading and grasping what's written. Not performing the task.

tahliar commented on articles/network-bond-managing-devices.asm.xml:
```diff
@@ -0,0 +1,144 @@
+          <term>GOAL</term>
+            <listitem>
+              <para>
+                What's the reader's take-away from this article?
```
Seems a bit redundant tbh

janajaeger replied:
I agree and if you decide not to have this, it's totally fine. It all depends on the wording you choose for the "What" section. If that intro covers the goal, then don't repeat it.

tahliar commented:
I'll probably re-test these procedures (I tested the configuration topics for the previous smart docs but it can't hurt to do it again), but otherwise this is ready for feedback on structure, title choices, etc.

tahliar commented on articles/network-team-managing-devices.asm.xml:
```diff
@@ -0,0 +1,152 @@
+      <meta name="updated" content="2023-06-23" its:translate="no"/>
+      <meta name="architecture" content="x86;power" its:translate="no"/>
+      <meta name="productname" its:translate="no">
+        <productname version="1">&alp;</productname>
```
@fsundermeyer @taroth21 @janajaeger 
ALP uses NetworkManager instead of Wicked, doesn't it? But this article includes some Wicked steps. Since we don't have a version-switcher currently, should I just remove ALP from the productname list and keep the Wicked steps for now? How should I handle this?

taroth21 replied:
Hi Tahlia, please find my feedback in the two annotated PDFs below. The first one holds some comments that are valid for the second one as well, as they share some content. 

[network-bond-managing-devices_feedback_taroth.pdf](https://github.com/SUSE/doc-modular/files/12303358/network-bond-managing-devices_feedback_taroth.pdf)

[network-team-managing-devices_feedback_taroth.pdf](https://github.com/SUSE/doc-modular/files/12303362/network-team-managing-devices_feedback_taroth.pdf)

I would suggest to discuss your latest question above in our next sync call and will put in on the agenda.

tahliar replied:
Thanks Tanja, I've applied your feedback and will update the PDFs in the description with the updated content.

For the intro paragraphs in the comparison topic, I decided to just remove all the distracting content and stick to the basic facts. Let me know what you think :)

taroth21 reviewed (changes requested):
See the annotated PDFs.

tahliar commented:
@Firstyear can you please review?

Firstyear commented on articles/network-bond-managing-devices.asm.xml:
```diff
@@ -0,0 +1,148 @@
+          <term>WHAT?</term>
+          <listitem>
+            <para>
+              Combine two or more network cards into a single bonding device.
```
interfaces (not cards) here. For example, there can be enterprise network cards that have multiple interfaces on them.

Firstyear commented on articles/network-bond-managing-devices.asm.xml:
```diff
@@ -0,0 +1,148 @@
+          <term>WHY?</term>
+          <listitem>
+            <para>
+              Network bonding can increase bandwidth or provide redundancy.
```
and/or rather than or?

Firstyear commented on articles/network-bond-managing-devices.asm.xml:
```diff
@@ -0,0 +1,148 @@
+              </listitem>
+              <listitem>
+                <para>
+                  Existing network cards to include in the bonding device
```
As above, interfaces not cards.

Firstyear commented on references/network-bond-team-comparison.xml:
```diff
@@ -0,0 +1,133 @@
+    <abstract>
+      <para>
+       <emphasis>Network bonding</emphasis> and <emphasis>network teaming</emphasis>
+       are different methods for combining network connections to provide a logical layer.
```
"to provide a single combined interface". Maybe better wording of it?

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+    <meta name="maintainer" content="tahlia.richardson@suse.com" its:translate="no"/>
+    <abstract>
+      <para>
+        Network bonding combines two or more network cards into a single bonding device to increase
```
"single bonded device".

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+      </listitem>
+      <listitem>
+        <para>
+          Existing network cards to include in the bonding device
```
interfaces

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+    <tip>
+      <title>Bonding and &xen;</title>
+      <para>
+        Bonding devices are made up of multiple real network cards. In most configurations, this
```
interfaces.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+      <para>
+        Bonding devices are made up of multiple real network cards. In most configurations, this
+        means that you should only set up bonding in &dom0;. To set up bonding in a &vmguest;,
+        you must have multiple network cards assigned to the &vmguest; system.
```
I need to think about this line, there is something I want to say about how you have to configure dom0 here, but I can't quite connect the words right now.

Firstyear replied:
Bonding devices are made up of multiple network interfaces. In most configurations you should configure bonding in dom0. Virtual interfaces to guests are then created as a bridge with these bonded devices. This simplifies your guest creation and deployment. It is possible but not advised to configure bonding in a guest. When configuring bonding in a guest you must assign multiple interfaces to the guest, and configure dom0 *without* bonding (else you get bonding inside bonding). You must also be careful with how you configure dom0 and it's network bridges that you do not mix bonding inside dom0 and inside guests.

tahliar replied:
Currently this note is only included in the bonding article but I imagine it's also relevant for teaming so I might add it to that article as well.

tahliar replied:
I don't know why I've only just thought of this, but surely this applies to KVM as well? Maybe we should make this note apply more broadly by saying that you should usually set up bonding on the host, not the guest? (I'm not as familiar with Xen so correct me if that's not what Dom0 is).

Firstyear replied:
You're absolutely correct, dom0 is the host on xen, similar to a host in kvm. And yes, this applies equally to both, where you should configure these on the host, not the vm.

Generally there are lots of reasons for this, but a big one is that virtual ethernet devices in kvm, even if they say things like "gigabit" they actually run at "infinite" speed, limited by memory bandwidth iirc. So you don't need to link aggregate in a vm, because the vm can consume the full bandwidth internally, and the host only needs the link agg to the external world. But also it's just simpler. If you have a huge vm host running 40 vms, you'd need to configure link aggs on all the vms, and I think it also would affect the ability to do guest migrations then too. Where as doing it on the host, means you configure it once. makes it all much easier.

tahliar replied:
Great, thanks! How about: 

> **Bonding and virtualization**
> 
> Bonded devices are made up of multiple network interfaces. In most configurations you should only configure bonding in the host. Virtual interfaces to guests are then created as a bridge with the bonded devices, simplifying guest creation and deployment.
> 
> It is possible, but not recommended, to configure bonding in a guest. When configuring bonding in a guest you must assign multiple interfaces to the guest, and configure the host without bonding. You must also be careful to configure the host and its network bridges so that you do not mix bonding in the host and guests.

Firstyear replied:
Perfect! Looks great :)

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+  <section xml:id="network-bond-modes">
+    <title>Bonding modes</title>
+    <para>
+      The following bonding modes are available:
```
It's worth pointing out in the bonding modes, that sometimes switch support is or is not needed for this to work. In some cases as well, it can also cause issues with some switches too. So I'm going to try to call out these cases.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        <listitem>
+          <para>
+            Packets are transmitted in round-robin fashion from the first to the last available
+            interface. Provides fault tolerance and load balancing.
```
Requires switch support. Some switches may fail/reject with this mode.

tahliar replied:
Is there anything we can point to for people to find out if their switch doesn't work with this mode? Or is it just the classic "check your hardware manual"?

Firstyear replied:
Pretty much a check your manual situation. You need to create an "etherchannel" which is where on the switch multiple ports act as a single port for the purposes of "what mac address is at the other end". 

The way that the "non switch" dependent modes work is:

* active-passive - one interface is "up" and the other is "down". If the up interface drops, the "down" interface goes up. This causes the switch to update it's mac address tables / arp tables to point to the second interface rather than the former primary.
* mode 5/6 (adaptive). All incoming traffic comes through "one" interface but outgoing traffic is balanced across *both* interfaces. This way the switch thinks everything goes to one mac address. The switch doesn't need support because on the second interface it just sees frames destined for another host and forwards them, but responses come back through the primary interface. You can think of this as "triangular ethernet" in a way. 

All the other modes require switches to actively be aware that two interfaces on the switch will be sharing a common mac address, and how to distribute traffic between those ports. Normally if you present the same mac on two switch ports that are not bound together (etherchannel in cisco terminology), the switch can be confused about where to actually send the data, and in some cases will disable the port, send the traffic to both ports, or just do something chaotic and unpredictable. 

When the switch is configured correctly though, it knows that an incoming frame for some MAC can go through port A or B, and there will be a load balancing algo to determine this. LACP is preferred here because first, it can dynamically tell the switch which ports are active or disabled in the "etherchannel" at a time. LACP can also dynamically change the load balancing algorithm in some cases to prevent link flooding (for example xor mode is vulnerable to this where two clients can hash to the same link in the bond and so will flood one link while the other is unused).

So yes, in the "switch support required" it will be "read your manual". Generally any l2 (or greater) managed switch will be capable of the required support. Generally "unmanaged" switches are the only ones that lack the ability to do these modes.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        <term><guimenu>1</guimenu> (<guimenu>active-backup</guimenu>)</term>
+        <listitem>
+          <para>
+            Only one network interface is active. If it fails, a different interface becomes active.
```
No switch support needed. No issues with switches.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+          <para>
+            Traffic is split between all available interfaces based on the number of interfaces
+            included in the bonding device. Requires support from the switch. Provides
+            fault tolerance and load balancing.
```
Requires switch support. Some switches may fail/reject with this mode.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        <listitem>
+          <para>
+            All traffic is broadcast on all interfaces. Requires support from the switch.
+            Provides fault tolerance.
```
Switch support required. Some switches may fail/reject with this mode.

If possible, use mode 1, or use this mode to provide sniffing capability by connecting each member of the bond to different switches / devices.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><guimenu>4</guimenu> (<guimenu>802.3ad</guimenu>)</term>
```
Also called LACP.

Requires switch support for LACP. Not all switches support this.

This is the *preferred mode* to use if your switch supports it. It will give you the least headaches!

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        <term><guimenu>4</guimenu> (<guimenu>802.3ad</guimenu>)</term>
+        <listitem>
+          <para>
+            Aggregates interfaces into groups that share the same speed and duplex settings.
```
More that each interface in the LACP group must be identical WRT to speed/duplex settings. Other modes like active-backup (mode 1) can have mixed mode interfaces like a gigabit port and a 10/100 fallback. But LACP requires all members to be the same speed/duplex, and must be connected to the same switch.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><guimenu>5</guimenu> (<guimenu>balance-tlb</guimenu>)</term>
```
No switch support needed. Some switches may fail/reject with this mode.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><guimenu>6</guimenu> (<guimenu>balance-alb</guimenu>)</term>
```
No switch support needed. Some switches may fail/reject with this mode.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,252 @@
+  </section>
+  <section xml:id="network-bond-configure-with-yast">
+    <title>Configuring network bonding with &yast;</title>
+<!--trichardson 2023-05-19: It might be good to also add a section for configuring bonding manually?-->
```
Most admins will want to configure manually, not with yast. So yes, manual config should be docced.

tahliar replied:
Thanks, I'll work on adding a new section for this.

Firstyear replied:
I'd be happy to send through some configs, I think I've done this before with the kernel lacp setup before. It's honestly not that hard. You could also just do the yast setup, then check what it spits out into /etc/sysconfig/network/ifcfg*

tahliar replied:
Oh thanks, good idea, I've probably still got the ifcfg files on my test VMs (assuming I didn't do any weird tests after that to mess them up). I'm assuming it will be pretty similar to the teaming procedure, at least for the stuff you do after creating the ifcfg file.

Firstyear replied:
Pretty much. Just make the ifcfg files and reboot is pretty much what  Idid 😅

tahliar replied:
Okay so story time: when I originally tested the teaming procedure (where it says to remove the ifcfg files for the member interfaces), I removed the line "Wicked will re-create the configuration files with the necessary parameters for teaming", because I found that this was not true (unless it saves them somewhere else?). Everything still seemed to work, so I thought nothing more of it. 

But now that I'm testing manual config for bonding, I realise that you kinda do need those files for the hotplugging procedure that comes after this one. I tested creating the bond and removing the eth0/1 files, and again, wicked did not recreate them. I also tested creating the bond and then doing nothing to the eth0/1 files, and wicked updated them to change STARTMODE to "hotplug" but did not change BOOTPROTO to "none". 

So I guess my question is, would it be better to tell people to manually change the bond member config files to STARTMODE=hotplug and BOOTPROTO=none, rather than relying on wicked to do it?

Firstyear replied:
Yeah, I think if people are doing it manually, we should advise where they need to change and manually edit these files, rather than relying on wicked. I think most people wanting to do it manually will want "absolute control" especially with config management tools like ansible, so they'll want to create and manage all the files.

So I think we should tell them what to do, rather than rely on wicked :)

Firstyear commented on tasks/network-bond-enable-hotplug.xml:
```diff
@@ -0,0 +1,103 @@
+    <meta name="maintainer" content="tahlia.richardson@suse.com" its:translate="no"/>
+    <abstract>
+      <para>
+        In some network environments (such as &ha;), you might need to replace a bond port
```
bond member (not bond port)

Firstyear commented on tasks/network-bond-enable-hotplug.xml:
```diff
@@ -0,0 +1,103 @@
+    <abstract>
+      <para>
+        In some network environments (such as &ha;), you might need to replace a bond port
+        interface with a new one, for example, if a network device is constantly failing.
```
"if a network interface has a fault". Could be better wording.

Firstyear commented on tasks/network-team-configure.xml:
```diff
@@ -0,0 +1,306 @@
+        <term><guimenu>roundrobin</guimenu></term>
+        <listitem>
+          <para>
+            Packets are transmitted in round-robin fashion from the first to the last available
```
I think this requires switch support.

Firstyear commented on tasks/network-team-configure.xml:
```diff
@@ -0,0 +1,306 @@
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><guimenu>loadbalance</guimenu></term>
```
Doesn't require switch support.

Firstyear commented on tasks/network-team-configure.xml:
```diff
@@ -0,0 +1,306 @@
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><guimenu>lacp</guimenu></term>
```
This is always the preferred option.

Firstyear commented on tasks/network-team-configure.xml:
```diff
@@ -0,0 +1,306 @@
+IPADDRESS="&subnetI;.1/24"
+IPADDR6="fd00:deca:fbad:50::1/64"
+
+TEAM_RUNNER="loadbalance"
```
Should show lacp or activebackup as default here.

tahliar replied:
The example here is titled "Configuration for load balancing" so I dunno about this one. But if you think a different example would be better here, I'm open to changing it entirely.

Firstyear replied:
lacp is the better choice then :) it's what 99% of deployments will be wanting to use in enterprise environments. So lets demonstrate what people use and is best practice. :)

tahliar replied:
Will the lacp config look much different from the activebackup config (Example 2)? I guess the original idea behind having the loadbalance example was that it has those additional TEAM_LB parameters.

Firstyear replied:
Honestly, not having used teamd, I am not sure. I still think we should demonstrate LACP mode since that's what mostn people will want, and the TEAM_LB params may not be hugely important for everyone to actually need to overwrite anyway. Alternately, two examples?

tahliar replied:
There's already two examples - load balancing and fail over (activebackup). I could move the activebackup example to be first, if it's more common. Also happy to add a LACP example, although maybe if the config isn't very different from the activebackup config we could change that example to use LACP instead.

Firstyear replied:
I think I'm thinking about this as a sysadmin. If I was an admin I'd either have:

* A managed switch, and want LACP
* Or I have an unmanaged switch, so I want active-backup

That's why I'm thinking the load balance example isn't really helping anyone here, it's not a common setup that will be used or sought after. So I think we should prioritise what people will want to use/find in our docs :)

tahliar replied:
Ok cool, no worries, we can concentrate on activebackup and lacp. So would their configs look different enough that we'd want to put an example for each one? Or are they basically the same except for TEAM_RUNNER (in which case let's just have one example, and it can be lacp no trouble)?

Firstyear replied:
I think that seems reasonable to me. I think if they're really similar we just do LACP then :)

tahliar replied:
Had a look in `man ifcfg-team` and lacp was more similar to loadbalance than to activebackup, it turns out, so I adjusted the loadbalance example to be for lacp instead. Hopefully it looks okay.

Firstyear replied:
Looks fantastic! Thank you!

tahliar commented:
@Firstyear Thanks for your input. I've updated with most of your feedback. Items I still need to update are the virtualisation note, the teaming examples, and an ifcfg section for bonding.

tahliar commented:
Updated the virtualisation note and the teaming examples. Still working on bonding via ifcfg.

Firstyear commented:
> Updated the virtualisation note and the teaming examples. Still working on bonding via ifcfg.

you're doing great work here :) lmk when you're done with the bonding ifcfg files and I'll happily review.

tahliar commented:
@Firstyear I've now added a procedure for bonding with ifcfg. I couldn't find any info to confirm whether the same member interface config applies to teaming, so I left that procedure as it is. 

I also adjusted the procedure for removing a teamed device so that it's applicable to both and added it to the bond article, and made some minor improvements to the VLAN procedure.

Firstyear reviewed (approved):
Looks awesome! Thanks for your patience with me and all my requests here 😅

Firstyear commented:
Latest changes look good to me :)

tahliar commented:
For the sake of the impending localisation deadline, I'd prefer to add a good troubleshooting section later, rather than rush out something mediocre. Otherwise, this is ready for final reviews.

Firstyear commented on tasks/network-bond-configure.xml:
```diff
@@ -53,6 +53,16 @@
+          <para>
+            In most hardware setups, all network interfaces in a bonded device must be connected to
+            the same switch. Splitting a bond over multiple switches is only possible with specific
+            switches and careful configuration.
```
You should just say "consult your switch vendor docs" or similar here, otherwise looks good :)

tahliar commented on tasks/network-bond-configure.xml:
```diff
@@ -0,0 +1,449 @@
+    </abstract>
+  </info>
+  <section xml:id="network-bond-requirements">
+    <title>Requirements</title>
```
True, but if this topic were ever re-used anywhere it would need to include the requirements.
